contracts/FeralfileArtworkV4_2.sol

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import {Nonces} from "./Nonces.sol";
+import {FeralfileExhibitionV4_1} from "./FeralfileArtworkV4_1.sol";
+import {IFeralfileVaultV2} from "./IFeralfileVaultV2.sol";
+import {FeralfileSaleDataV2} from "./FeralfileSaleDataV2.sol";
+
+contract FeralfileExhibitionV4_2 is
+    FeralfileExhibitionV4_1,
+    FeralfileSaleDataV2,
+    Nonces
+{
+    error TokenIDNotFound();
+    error FunctionNotSupported();
+    error SaleNotStarted();
+    error InvalidPaymentAmount();
+    error TotalBpsOver();
+    error InvalidAddress();
+
+    // vault contract instance
+    IFeralfileVaultV2 public vaultV2;
+
+    mapping(uint256 => uint256) private seriesNextPurchasableTokenIds; // seriesID -> tokenID
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        bool burnable_,
+        bool bridgeable_,
+        address signer_,
+        address vault_,
+        address costReceiver_,
+        string memory contractURI_,
+        uint256[] memory seriesIds_,
+        uint256[] memory seriesMaxSupplies_,
+        uint256[] memory seriesNextPurchasableTokenIds_
+    )
+        FeralfileExhibitionV4_1(
+            name_,
+            symbol_,
+            burnable_,
+            bridgeable_,
+            signer_,
+            vault_,
+            costReceiver_,
+            contractURI_,
+            seriesIds_,
+            seriesMaxSupplies_
+        )
+    {
+        vaultV2 = IFeralfileVaultV2(payable(vault_));
+        for (uint256 i = 0; i < seriesIds_.length; i++) {
+            seriesNextPurchasableTokenIds[seriesIds_[i]] = seriesNextPurchasableTokenIds_[i];
+        }
+    }
+
+    /// @notice Set vaultV2 contract
+    /// @dev don't allow to set vaultV2 as zero address
+    function setVaultV2(address vault_) external onlyOwner {
+        if (vault_ == address(0)) {
+            revert InvalidAddress();
+        }
+
+        vaultV2 = IFeralfileVaultV2(payable(vault_));
+    }
+
+    /// @notice pay to get artworks to a destination address. The pricing, costs and other details is included in the saleData
+    /// @param r_ - part of signature for validating parameters integrity
+    /// @param s_ - part of signature for validating parameters integrity
+    /// @param v_ - part of signature for validating parameters integrity
+    /// @param saleData_ - the sale data
+    function buyBulkArtworks(
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        SaleDataV2 calldata saleData_
+    ) external payable {
+        if (!_selling) {
+            revert SaleNotStarted();
+        }
+
+        super._checkContractOwnedToken();
```

Comment:
This is not enough for checking the available token for sale. We should compare with the sale data quantity.